presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeWindows.java

```diff
@@ -85,30 +85,45 @@ protected PlanNode visitPlan(
                 RewriteContext<Multimap<WindowNode.Specification, WindowNode>> context)
```

Comment:
I wanted to issue a separate PR for optimal solution.
I don't think I understand how your algorithm would solve the problem optimally. It looks like a greedy approach to me.
I believe the problem we have (minimize the number of subsets of independent nodes) here is NP-hard. I can reduce it to graph coloring problem (on subset of graphs only though), which proves it's NP-hard.
However, here we don't deal with long plans anyway, so I was thinking about providing an optimal solution (in exponential complexity) with the cap on the size of the plan, in case of rare cases when the plan is extremely big (then we'd fall back to the greedy approach I suggested in this PR).
This would be a separate PR, if needed, though.